### PR TITLE
fix(eslint-plugin): [prefer-function-type] disable fix when using comments

### LINKED ISSUE #3004: [prefer-function-type] doc comment deleted when fixed
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-function-type": ["error"]
  }
}
```

```TS
interface X {
  /** @deprecated */
  (): void;
}
```

**Expected Result**

```ts
type X =
  /** @deprecated */
  () => void;
```

**Actual Result**

```ts
type X = () => void;
```

**Additional Info**

Maybe it'd be nice to also have an option to prevent transform when a doc comment is present.
I didn't check if regular comments get eaten too.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.14.2` |
| `@typescript-eslint/parser`        | `4.14.2` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.19.0` |
| `node`                             | `14.15.4` |

### PR DESCRIPTION
**Pull request by arminyahya**
Fixed #3004

### CODE REVIEW

**Comment by bradzacher** [comment]
I feel like we could probably handle some simple comment cases.
Eg if there's exactly 1 comment, it'd be pretty trivial for us to move it above when fixing:

```ts
interface Foo1 {
  // comment
  (): void;
}
// comment
type Foo1 = () => void;

// ---

interface Foo2 {
  /* comment */
  (): void;
}
/* comment */
type Foo2 = () => void;

// ---

interface Foo3 {
  /**
   * comment
   */
  (): void
}
/**
   * comment
   */
type Foo3 = () => void;

// or... could correct the indent as well
/**
 * comment
 */
type Foo3 = () => void;
```

We can disable the fixer for more complex cases, but it isn't hard for us to handle *at least* the first two cases of one, single-line comment before the type.

WDYT?

**Comment by arminyahya** [comment]
I agree with you. 
I make a change in a few days.

**Comment by arminyahya** [comment]
I agree with you. 
I make a change in a few days.

**Comment by arminyahya** [comment]
@bradzacher 
looks like we can achieve all cases you mentioned.
what do you suggest for formatting comments?

`import { format } from 'prettier';` ?

**Comment by bradzacher** [comment]
We can't use prettier for various reasons.

Instead you'd have to split the comment by lines and cut the extra whitespace from each line (other than the first).

Youd figure out how much whitespace exists by examining the source code and figuring out the indent.

I'm more than okay with not handling this case right now, as it's really non-trivial. We can always come back and add it in a separate PR.

File: packages/eslint-plugin/src/rules/prefer-function-type.ts [hunk 1]
```diff
@@ -141,33 +101,91 @@ export default util.createRule({
...
+                let commentText =
+                  comment.type === AST_TOKEN_TYPES.Line
+                    ? `//${comment.value}`
+                    : `/*${comment.value}*/`;
+                const isCommentOnTheSameLine =
+                  comment.loc.start.line === member.loc.start.line;
+                if (!isCommentOnTheSameLine) {
+                  commentText += '\n';
+                } else {
+                  commentText += ' ';
```
Comment (arminyahya) [inline packages/eslint-plugin/src/rules/prefer-function-type.ts:177]: I'm not sure about this line. WDYT?  @bradzacher

**Comment by arminyahya** [comment]
I hope my **rebase** not make it difficult to review.
Looks like i have to increase my codecov to reach 90%.I will do it after we review prefer-function-type changes.
these changes fix the **first three cases** you mentioned.

**Comment by armano2** [comment]
i think that your rebase is not correct, you should rebase it over current master not some old revision

**Comment by bradzacher** [comment]
It looks like you still need to rebase your changes.

Don't be afraid to cherry pick your commits to a new branch and force push overwrite your remote branch.

**Comment by arminyahya** [comment]
Let me just close this and open another PR from another branch. I will refer to this.

**Pull request closed without merging**